truffle/src/main/ruby/core/proc.rb

```diff
@@ -33,18 +33,41 @@ class Proc
 
   def curry(curried_arity = nil)
     if lambda? && curried_arity
+
       if arity >= 0 && curried_arity != arity
         raise ArgumentError, "Wrong number of arguments (%i for %i)" % [
           curried_arity,
           arity
         ]
       end
 
-      if arity < 0 && curried_arity < (-arity - 1)
-        raise ArgumentError, "Wrong number of arguments (%i for %i)" % [
-          curried_arity,
-          -arity - 1
-        ]
+      if arity < 0
+        total_args = 0
+        req_args = 0
+        has_rest = false
+
+        for p in parameters do
```

Comment:
Is this code from Rubinius or did you write this yourself? `for ... in ... do` is antiquated Ruby syntax. We'd normally write `parameters.each do |p|`. The only difference (beyond opinion of style) is that the latter creates a new scope. If that matters, put a nil declaration of the variables you need in the upper scope before the loop.